### PR TITLE
*: use structure to store user identity

### LINKED ISSUE #3740: Cannot authenticate users which username contains '@'
MySQL:

```
mysql> create user 'test@example.com'@'localhost' identified by '';
Query OK, 0 rows affected (0.03 sec)
```

```
$ mysql -u 'test@example.com' --host 127.0.0.1
Welcome to the MySQL monitor.  Commands end with ; or \g.
....
```

TiDB:

```
mysql> create user 'test@example.com'@'localhost' identified by '';
Query OK, 1 row affected (0.00 sec)
```

```
$ mysql -u 'test@example.com' --host 127.0.0.1 --port 4000
ERROR 1045 (28000): Access denied for user 'test@example.com'@'127.0.0.1' (using password: YES)
```

### PR DESCRIPTION
**Pull request by breeswish**
Fix #3740 

@zimulala @tiancaiamao PTAL

### CODE REVIEW

File: privilege/privileges/cache.go [hunk 1]
```diff
@@ -350,6 +350,7 @@ func patternMatch(str string, patChars, patTypes []byte) bool {
 func (p *MySQLPrivilege) connectionVerification(user, host string) *userRecord {
 	for i := 0; i < len(p.User); i++ {
 		record := &p.User[i]
+		fmt.Printf("record #%d: %v    (user = %s, host = %s)\n", i, record, user, host)
```
Comment (tiancaiamao) [inline privilege/privileges/cache.go]: remove debug log

Comment (tiancaiamao) [inline privilege/privileges/cache.go]: rest looks good to me.

**Comment by tiancaiamao** [comment]
LGTM

**Comment by zz-jason** [comment]
@breeswish plz resolve conflicts

**Comment by breeswish** [comment]
@zz-jason Fixed

**Pull request merged**